### PR TITLE
[react-interactions] More Tab Focus control handling

### PR DESCRIPTION
**Pull request by trueadm**
This is a follow up PR to the previous initial work to handle basic tab focus control with `TabFocusController.js`s.

This PR adds a few more features to the experimental scope API and, most importantly, provides an internal imperative mechanism for controlling tab focus via an API. I've also made containment an optional prop on `TabFocusController`s and provided a way of exposing an imperative ref to scopes so these can easily be handed back via scope traversal.

In my next PR I plan on demonstrating taking this future and building components that allow for a powerful Grid based keyboard focusable system (using arrows + tab + modifiers).

### CODE REVIEW

File: packages/react-dom/src/client/focus/TabFocusController.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  getNextController: () => null | ControllerHandle,
+  getPreviousController: () => null | ControllerHandle,
+|};
+
+const {useImperativeHandle, useRef} = React;
+
+export const TabFocusController = React.forwardRef(
+  ({children, contain}: TabFocusControllerProps, ref): React.Node => {
+    const scopeRef = useRef(null);
+    const keyboard = useKeyboard({onKeyDown, preventKeys: ['tab']});
```
Comment (necolas) [inline packages/react-dom/src/client/focus/TabFocusController.js]: tab -> Tab

File: packages/react-dom/src/client/focus/TabFocusController.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      focusFirst,
+      focusNext,
+      focusPrevious,
+      getNextController,
+      getPreviousController,
+    };
+
+    useImperativeHandle(ref, () => controllerHandle);
+
+    function onKeyDown(event: KeyboardEvent): boolean {
```
Comment (necolas) [inline packages/react-dom/src/client/focus/TabFocusController.js]: Let's move this into the useKeyboard call

File: packages/react-dom/src/client/focus/TabFocusController.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    const controllerHandle: ControllerHandle = {
+      focusFirst,
+      focusNext,
+      focusPrevious,
+      getNextController,
+      getPreviousController,
+    };
+
+    useImperativeHandle(ref, () => controllerHandle);
```
Comment (necolas) [inline packages/react-dom/src/client/focus/TabFocusController.js]: Note this hook is not good for perf according to my tests

Comment (trueadm) [reply packages/react-dom/src/client/focus/TabFocusController.js]: No other way to do what we want here, other than using a class. :/

File: packages/react-dom/src/client/focus/TabFocusController.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+type KeyboardEventType = 'keydown' | 'keyup';
+
+type KeyboardEvent = {|
+  altKey: boolean,
+  ctrlKey: boolean,
+  isComposing: boolean,
+  key: string,
+  location: number,
+  metaKey: boolean,
+  repeat: boolean,
```
Comment (necolas) [inline packages/react-dom/src/client/focus/TabFocusController.js]: AFAIK, `repeat` is unreliable. We could ditch `location`, `repeat`, and maybe even `isComposing` from these events

File: packages/react-dom/src/client/focus/TabFocusController.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        if (event.key !== 'Tab') {
+          return true;
+        }
+        if (event.shiftKey) {
+          return focusPrevious();
+        } else {
+          return focusNext();
+        }
+      },
+      preventKeys: ['Tab'],
```
Comment (necolas) [inline packages/react-dom/src/client/focus/TabFocusController.js]: I think you probably also need `[ 'Tab', { shiftKey: true } ]`

File: packages/react-dom/src/client/focus/TabFocusController.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    function getPreviousController(): null | ControllerHandle {
+      const tabbableScope = scopeRef.current;
+      const allScopes = tabbableScope.getChildrenFromRoot();
+      if (allScopes === null) {
+        return null;
+      }
+      const currentScopeIndex = allScopes.indexOf(tabbableScope);
+      if (currentScopeIndex === -1 || currentScopeIndex === 0) {
```
Comment (necolas) [inline packages/react-dom/src/client/focus/TabFocusController.js]: `<= 0`?

File: packages/react-dom/src/client/focus/__tests__/TabFocusController-test.internal.js [hunk 7]
```diff
@@ -221,5 +215,44 @@ describe('TabFocusContainer', () => {
...
+          </TabFocusController>
+          <TabFocusController ref={secondFocusControllerRef}>
+            <input tabIndex={-1} />
+            <div ref={divRef} tabIndex={0} />
+          </TabFocusController>
+        </div>
+      );
+
+      ReactDOM.render(<Test />, container);
+      firstFocusControllerRef.current.focusFirst();
```
Comment (necolas) [inline packages/react-dom/src/client/focus/__tests__/TabFocusController-test.internal.js]: The tests are a bit easier to read if we avoid writing out `ref.current` each time:

```
const button = buttonRef.current;
const firstFocusController = firstFocusControllerRef.current;

firstFocusController.focusFirst();
expect(document.activeElement).toBe(button);
firstFocusController.focusNext();
expect(document.activeElement).toBe(button2);
```

**Pull request merged**